services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py

```diff
@@ -0,0 +1,259 @@
+    networks = [
+        x
+        for x in (await docker_client.networks.list())
+        if "swarm" in x["Scope"] and network_name == x["Name"]
+    ]
+    logger.debug(f"found the following swarm networks: {networks=}")
+    if not networks:
+        raise ValueError(f"network {network_name} not found")
+    if len(networks) > 1:
+        raise ValueError(
+            f"network {network_name} is ambiguous, too many network founds: {networks=}"
+        )
+    logger.debug("found a network %s", f"{networks[0]=}")
+    assert "Id" in networks[0]  # nosec
+    assert isinstance(networks[0]["Id"], str)  # nosec
+    return networks[0]["Id"]
+
+
+def create_service_config(
+    settings: AppSettings,
+    service_env: Dict[str, Any],
+    service_name: str,
+    network_id: str,
+    service_secrets: List[DockerSecret],
+    cmd: Optional[List[str]],
+    labels: Dict[str, str],
+) -> Dict[str, Any]:
+    env = deepcopy(service_env)
+    env.pop("PATH", None)
+    # create the secrets array containing the TLS cert/key pair
+    container_secrets = []
+    for s in service_secrets:
+        container_secrets.append(
+            {
+                "SecretName": s.secret_name,
+                "SecretID": s.secret_id,
+                "File": {
+                    "Name": f"{_DASK_KEY_CERT_PATH_IN_SIDECAR / s.secret_file_name}",
+                    "UID": "0",
+                    "GID": "0",
+                    "Mode": 0x777,
+                },
+            }
+        )
+        env_updates = {}
+        for env_name, env_value in env.items():
+            if env_value == s.secret_file_name:
+                env_updates[
+                    env_name
+                ] = f"{_DASK_KEY_CERT_PATH_IN_SIDECAR / s.secret_file_name}"
+        env.update(env_updates)
+    mounts = [
+        # docker socket needed to use the docker api
+        {
+            "Source": "/var/run/docker.sock",
+            "Target": "/var/run/docker.sock",
+            "Type": "bind",
+            "ReadOnly": True,
+        },
+        # the sidecar data data is stored in a volume
+        {
+            "Source": settings.COMPUTATIONAL_SIDECAR_VOLUME_NAME,
+            "Target": _SHARED_COMPUTATIONAL_FOLDER_IN_SIDECAR,
+            "Type": "volume",
+            "ReadOnly": False,
+        },
+    ]
+
+    container_config = {
+        "Env": env,
+        "Image": settings.COMPUTATIONAL_SIDECAR_IMAGE,
+        "Init": True,
+        "Mounts": mounts,
+        "Secrets": container_secrets,
+        # "Command": ["ls", "-tlah", f"{_DASK_KEY_CERT_PATH_IN_SIDECAR}"],
+    }
+    if cmd:
+        container_config["Command"] = cmd
+    return {
+        "name": service_name,
+        "labels": labels,
+        "task_template": {
+            "ContainerSpec": container_config,
+            "RestartPolicy": {"Condition": "on-failure"},
+        },
+        "networks": [network_id],
+        # "mode": {"Global": {}},
```

Comment:
When the client scales up, we know how much cpus/ram he wants. Can we use that for resource requiremnts and is there a way to let him know that we cannot start a worker any more? We could use that also for a trigger to scale the hardware.

Reply:
hmm I'm not sure we know that. actually there are 3 ways:
- auto-scaling where I am not sure how the gateway decides when to scale
- manual scaling from the client. how does the client know he can scale more?
- global services, then the scaling is done by the hardware scaler